DexPriceAggregatorUniswapV3/contracts/DexPriceAggregatorUniswapV3.sol

```diff
@@ -0,0 +1,409 @@
+    function assetToAsset(
+        address _tokenIn,
+        uint256 _amountIn,
+        address _tokenOut,
+        uint256 _twapPeriod
+    ) external view override returns (uint256 amountOut) {
+        if (_tokenIn == weth) {
+            return ethToAsset(_amountIn, _tokenOut, _twapPeriod);
+        } else if (_tokenOut == weth) {
+            return assetToEth(_tokenIn, _amountIn, _twapPeriod);
+        } else {
+            return
+                _fetchAmountCrossingPools(
+                    _tokenIn,
+                    _amountIn,
+                    _tokenOut,
+                    _twapPeriod
+                );
+        }
+    }
+
+    /// @notice Given a token and its amount, return the equivalent value in ETH
+    /// @param _tokenIn Address of an ERC20 token contract to be converted
+    /// @param _amountIn Amount of tokenIn to be converted
+    /// @param _twapPeriod Number of seconds in the past to consider for the TWAP rate
+    /// @return ethAmountOut Amount of ETH received for amountIn of tokenIn
+    function assetToEth(
+        address _tokenIn,
+        uint256 _amountIn,
+        uint256 _twapPeriod
+    ) public view returns (uint256 ethAmountOut) {
+        address tokenOut = weth;
+        address pool = _getPoolForRoute(
+            PoolAddress.getPoolKey(_tokenIn, tokenOut, defaultPoolFee)
+        );
+        return
+            _fetchAmountFromSinglePool(
+                _tokenIn,
+                _amountIn,
+                tokenOut,
+                pool,
+                _twapPeriod
+            );
+    }
+
+    /// @notice Given an amount of ETH, return the equivalent value in another token
+    /// @param _ethAmountIn Amount of ETH to be converted
+    /// @param _tokenOut Address of an ERC20 token contract to convert into
+    /// @param _twapPeriod Number of seconds in the past to consider for the TWAP rate
+    /// @return amountOut Amount of tokenOut received for ethAmountIn of ETH
+    function ethToAsset(
+        uint256 _ethAmountIn,
+        address _tokenOut,
+        uint256 _twapPeriod
+    ) public view returns (uint256 amountOut) {
+        address tokenIn = weth;
+        address pool = _getPoolForRoute(
+            PoolAddress.getPoolKey(tokenIn, _tokenOut, defaultPoolFee)
+        );
+        return
+            _fetchAmountFromSinglePool(
+                tokenIn,
+                _ethAmountIn,
+                _tokenOut,
+                pool,
+                _twapPeriod
+            );
+    }
+
+    /************************
+     * Management functions *
+     ************************/
+
+    /// @notice Override the Uniswap V3 pool queried on a tokenA:tokenB route
+    /// @dev Override can be reset by using address(0) for _pool
+    /// @param _tokenA Address of an ERC20 token contract
+    /// @param _tokenB Address of another ERC20 token contract
+    /// @param _pool Address of a Uniswap V3 pool constructed with _tokenA and _tokenB
+    function setPoolForRoute(
```

Comment:
- _tokenA and _tokenB should not be the zero address
- _tokenA should not be equal _tokenB